drv: ft8xx: support multiple instances

hubertmis opened the pull request:
Improve the FT8xx display driver to support multiple driver instances.

The ft8xx_reference_api.h still follows the FT8xx programming guides, limiting the usage of the driver to one instance.

hubertmis commented:
@kartben, what should I do to receive any feedback on this PR?

kartben commented on samples/drivers/misc/ft800/src/main.c:
```diff
@@ -23,63 +24,67 @@
 }
 
 int main(void)
 {
```
unrelated to your changes but this code should first perform an `if device_is_ready(...)` check

kartben commented on include/zephyr/drivers/misc/ft8xx/ft8xx.h:
```diff
@@ -45,8 +46,11 @@ struct ft8xx_touch_transform {
+ * @param user_data Pointer to user data provided during callback registration
  */
-typedef void (*ft8xx_int_callback)(void);
+typedef void (*ft8xx_int_callback)(const struct device *dev, void *user_data);
```
adding support for user data seems to be unrelated to adding support for multiple instances. It would be better to split the changes in different commits

kartben reviewed (changes requested):
some comments
Also, you should add a note in doc/releases/migration-guide-4.1.rst about the changes you've been making to the API, i.e. indicate that all driver functions now take a device pointer since support for multiple instances has been added AND indicate that there is now also a new user_data parameter added to `ft8xx_register_int` and to the callback signature `ft8xx_int_callback`

hubertmis committed:
drv: ft8xx: support multiple instances

Improve the FT8xx display driver to support multiple driver instances.

The ft8xx_reference_api.h still follows the FT8xx programming guides,
limiting the usage of the driver to one instance.

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

hubertmis committed:
drv: ft8xx: user data in the callback function

The user of the ft8xx driver can register a callback function called
on a touch detection interrupt event. This function is extended to
include a user-defined pointer. The user-defined pointer helps to share
a callback with multiple driver instances.

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

hubertmis commented:
> some comments Also, you should add a note in doc/releases/migration-guide-4.1.rst about the changes you've been making to the API, i.e. indicate that all driver functions now take a device pointer since support for multiple instances has been added AND indicate that there is now also a new user_data parameter added to `ft8xx_register_int` and to the callback signature `ft8xx_int_callback`

I've split the pull request into two commits and added notes to the migration guide.

hubertmis commented:
> some comments Also, you should add a note in doc/releases/migration-guide-4.1.rst about the changes you've been making to the API, i.e. indicate that all driver functions now take a device pointer since support for multiple instances has been added AND indicate that there is now also a new user_data parameter added to `ft8xx_register_int` and to the callback signature `ft8xx_int_callback`

@kartben , are there any more changes you are requesting?

kartben reviewed (approved):
> @kartben , are there any more changes you are requesting?

@hubertmis No :) but in the future make sure to re-request a review from the GitHub UI as it's the best way to make sure reviewers know the PR is actually ready for another round of review.
Thank you!

fabiobaltieri commented on drivers/misc/ft8xx/ft8xx.c:
```diff
@@ -97,131 +71,167 @@ static int ft8xx_init(const struct device *dev)
+	.hsync0   = DT_INST_PROP(idx, hsync0),                                                     \
+	.hsync1   = DT_INST_PROP(idx, hsync1),                                                     \
+};                                                                                                 \
+static struct ft8xx_data ft8xx_##idx##_data = {                                                    \
+	.ft8xx_dev = NULL,                                                                         \
+	.irq_callback = NULL,                                                                      \
+	.irq_callback_ud = NULL,                                                                   \
+												   \
+	.spi = SPI_DT_SPEC_INST_GET(idx, SPI_WORD_SET(8) | SPI_OP_MODE_MASTER, 0),                 \
+	.irq_gpio = GPIO_DT_SPEC_INST_GET(idx, irq_gpios),                                         \
+};                                                                                                 \
```
the first three don't need to be initialized, just drop them, spi and irq_gpio can go in the config struct, up to you whether to change them in this PR or later or not, not a huge difference but it'd save some ram and some flash
